lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -2519,12 +2519,26 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
                     kind.getSelector());
     }
 
-    // We don't support protocols outside the top level of a file.
-    if (isa<ProtocolDecl>(NTD) &&
-        !NTD->getParent()->isModuleScopeContext()) {
-      NTD->diagnose(diag::unsupported_nested_protocol, NTD->getName());
-      NTD->setInvalid();
-      return;
+    if (isa<ProtocolDecl>(NTD)) {
+      if (NTD->getASTContext().LangOpts.hasFeature(Feature::NestedProtocols)) {
+        // Protocols may only be nested in non-generic contexts.
+        if (NTD->getParent()->isGenericContext()) {
+          if (isa<ProtocolDecl>(NTD->getParent())) {
+            NTD->diagnose(diag::unsupported_nested_protocol_in_protocol, NTD->getName());
```

Comment:
Do you also want this tailored diagnostic for the following cases?

```swift
protocol P {}

extension P {
  protocol Nested {}
}

extension P {
  func f() {
    protocol Nested {}
  }
}
```

Reply:
Yes, that's a good point. I added the tailored diagnostic because I don't think most Swift developers would think of a plain protocol as a "generic context".